File: backend/trip-service/internal/service/trip_service.go
package service

import (
	"context"
	"fmt"
	"time"

	"bus-booking/shared/ginext"
	"bus-booking/trip-service/internal/client"
	"bus-booking/trip-service/internal/constants"
	"bus-booking/trip-service/internal/model"
	"bus-booking/trip-service/internal/model/booking"
	"bus-booking/trip-service/internal/model/payment"
	"bus-booking/trip-service/internal/repository"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
)

type TripService interface {
	SearchTrips(ctx context.Context, req *model.TripSearchRequest) ([]model.TripDetail, int64, error)
	GetTripByID(ctx context.Context, req *model.GetTripByIDRequest, id uuid.UUID) (*model.Trip, error)
	ListTrips(ctx context.Context, req *model.ListTripsRequest) ([]model.Trip, int64, error)
	GetSeatAvailability(ctx context.Context, tripID uuid.UUID) (*model.SeatAvailabilityResponse, error)
	GetTripsByRouteAndDate(ctx context.Context, routeID uuid.UUID, departureDate time.Time) ([]model.Trip, error)
	GetCompletedTripsForReschedule(ctx context.Context) ([]model.Trip, error)

	CreateTrip(ctx context.Context, req *model.CreateTripRequest) (*model.Trip, error)
	UpdateTrip(ctx context.Context, id uuid.UUID, req *model.UpdateTripRequest) (*model.Trip, error)
	DeleteTrip(ctx context.Context, id uuid.UUID) error
	RescheduleTrip(ctx context.Context, id uuid.UUID, newDeparture, newArrival time.Time) error
	CancelTrip(ctx context.Context, id uuid.UUID) error
	ProcessTripStatusUpdates(ctx context.Context) error
}

type TripServiceImpl struct {
	tripRepo      repository.TripRepository
	routeRepo     repository.RouteRepository
	routeStopRepo repository.RouteStopRepository
	busRepo       repository.BusRepository
	seatRepo      repository.SeatRepository
	bookingClient client.BookingClient
	paymentClient client.PaymentClient
}

func NewTripService(
	tripRepo repository.TripRepository,
	routeRepo repository.RouteRepository,
	routeStopRepo repository.RouteStopRepository,
	busRepo repository.BusRepository,
	seatRepo repository.SeatRepository,
	bookingClient client.BookingClient,
	paymentClient client.PaymentClient,
) TripService {
	return &TripServiceImpl{
		tripRepo:      tripRepo,
		routeRepo:     routeRepo,
		routeStopRepo: routeStopRepo,
		busRepo:       busRepo,
		seatRepo:      seatRepo,
		bookingClient: bookingClient,
		paymentClient: paymentClient,
	}
}

func (s *TripServiceImpl) SearchTrips(ctx context.Context, req *model.TripSearchRequest) ([]model.TripDetail, int64, error) {
	trips, total, err := s.tripRepo.SearchTrips(ctx, req)
	if err != nil {
		log.Error().Err(err).Msg("Failed to search trips")
		return nil, 0, ginext.NewInternalServerError("failed to search trips")
	}

	return trips, total, nil
}

func (s *TripServiceImpl) GetTripByID(ctx context.Context, req *model.GetTripByIDRequest, id uuid.UUID) (*model.Trip, error) {
	trip, err := s.tripRepo.GetTripByID(ctx, req, id)
	if err != nil {
		return nil, ginext.NewInternalServerError("failed to get trip")
	}

	if req.SeatBookingStatus && req.PreloadBus && req.PreloadSeat && trip.Bus != nil && len(trip.Bus.Seats) > 0 {
		seatIDs := make([]uuid.UUID, len(trip.Bus.Seats))
		for i, seat := range trip.Bus.Seats {
			seatIDs[i] = seat.ID
		}

		seatStatuses, err := s.bookingClient.GetSeatStatus(ctx, trip.ID, seatIDs)
		if err != nil {
			log.Error().Err(err).Msg("Failed to check seat status from booking service")
			return nil, ginext.NewInternalServerError("failed to get seat status")
		}

		seatStatusMap := make(map[uuid.UUID]booking.SeatStatus)
		for _, status := range seatStatuses {
			seatStatusMap[status.SeatID] = status
		}

		for i, seat := range trip.Bus.Seats {
			if status, ok := seatStatusMap[seat.ID]; ok {
				trip.Bus.Seats[i].Status = &status
				continue
			}
			trip.Bus.Seats[i].Status = &booking.SeatStatus{
				SeatID:   seat.ID,
				IsBooked: false,
				IsLocked: false,
			}
		}
	}

	return trip, nil
}

func (s *TripServiceImpl) ListTrips(ctx context.Context, req *model.ListTripsRequest) ([]model.Trip, int64, error) {
	// If IDs provided, fetch specific trips (batch mode)
	if len(req.IDs) > 0 {
		trips, err := s.tripRepo.GetTripsByIDs(ctx, req.IDs)
		if err != nil {
			return nil, 0, ginext.NewInternalServerError("failed to list trips by IDs")
		}
		return trips, int64(len(trips)), nil
	}

	// Otherwise, use pagination
	trips, total, err := s.tripRepo.ListTrips(ctx, req.Page, req.PageSize)
	if err != nil {
		return nil, 0, ginext.NewInternalServerError("failed to list trips")
	}
	return trips, total, nil
}

func (s *TripServiceImpl) GetSeatAvailability(ctx context.Context, tripID uuid.UUID) (*model.SeatAvailabilityResponse, error) {
	req := &model.GetTripByIDRequest{
		PreloadBus:  true,
		PreloadSeat: true,
	}
	trip, err := s.tripRepo.GetTripByID(ctx, req, tripID)
	if err != nil {
		return nil, ginext.NewInternalServerError("trip not found")
	}

	seats, err := s.seatRepo.GetListByBusID(ctx, trip.BusID)
	if err != nil {
		return nil, ginext.NewInternalServerError("failed to get seats")
	}

	// TODO: Check seat status from booking service
	var seatAvailabilities []model.SeatAvailability
	availableCount := 0

	for _, seat := range seats {
		seatAvail := model.SeatAvailability{
			SeatID:      seat.ID,
			SeatNumber:  seat.SeatNumber,
			SeatType:    seat.SeatType,
			Price:       trip.BasePrice * seat.PriceMultiplier,
			IsAvailable: seat.IsAvailable, // TODO: Check from booking service
			Row:         seat.Row,
			Column:      seat.Column,
			Floor:       seat.Floor,
		}

		if seatAvail.IsAvailable {
			availableCount++
		}

		seatAvailabilities = append(seatAvailabilities, seatAvail)
	}

	return &model.SeatAvailabilityResponse{
		TripID:         tripID,
		AvailableSeats: availableCount,
		TotalSeats:     len(seats),
		SeatMap:        seatAvailabilities,
	}, nil
}

// GetTripsByRouteAndDate gets trips by route and departure date
func (s *TripServiceImpl) GetTripsByRouteAndDate(ctx context.Context, routeID uuid.UUID, departureDate time.Time) ([]model.Trip, error) {
	// Validate inputs
	if routeID == uuid.Nil {
		return nil, ginext.NewBadRequestError("route ID is required")
	}

	// Check if route exists
	_, err := s.routeRepo.GetRouteByID(ctx, routeID)
	if err != nil {
		return nil, fmt.Errorf("invalid route: %w", err)
	}

	// Get trips by route and date
	trips, err := s.tripRepo.GetTripsByRouteAndDate(ctx, routeID, departureDate)
	if err != nil {
		return nil, fmt.Errorf("failed to get trips: %w", err)
	}

	return trips, nil
}

func (s *TripServiceImpl) CreateTrip(ctx context.Context, req *model.CreateTripRequest) (*model.Trip, error) {
	if req.ArrivalTime.Before(req.DepartureTime) {
		return nil, ginext.NewBadRequestError("arrival time must be after departure time")
	}

	if req.DepartureTime.Before(time.Now()) {
		return nil, ginext.NewBadRequestError("departure time cannot be in the past")
	}

	// Check if route exists
	_, err := s.routeRepo.GetRouteByID(ctx, req.RouteID)
	if err != nil {
		return nil, ginext.NewBadRequestError("invalid route")
	}

	// Check if bus exists and is available
	bus, err := s.busRepo.GetBusByID(ctx, req.BusID)
	if err != nil {
		return nil, ginext.NewBadRequestError("invalid bus")
	}

	if !bus.IsActive {
		return nil, ginext.NewBadRequestError("bus is not active")
	}

	// Check for bus conflicts (same bus cannot have overlapping trips)
	conflictTrips, err := s.tripRepo.GetTripsByBusAndDateRange(ctx, req.BusID,
		req.DepartureTime.Add(-4*time.Hour), req.ArrivalTime.Add(4*time.Hour))
	if err != nil {
		return nil, ginext.NewInternalServerError("failed to check bus availability")
	}

	for _, existingTrip := range conflictTrips {
		if req.ArrivalTime.After(existingTrip.DepartureTime) && req.DepartureTime.Before(existingTrip.ArrivalTime) {
			return nil, ginext.NewBadRequestError("bus is already assigned to another trip during the specified time")
		}
	}

	trip := &model.Trip{
		RouteID:       req.RouteID,
		BusID:         req.BusID,
		DepartureTime: req.DepartureTime,
		ArrivalTime:   req.ArrivalTime,
		BasePrice:     req.BasePrice,
		Status:        "scheduled",
		IsActive:      true,
	}

	if err := s.tripRepo.CreateTrip(ctx, trip); err != nil {
		log.Error().Err(err).Msg("Failed to create trip")
		return nil, ginext.NewInternalServerError("failed to create trip")
	}

	// Load relationships
	return s.GetTripByID(ctx, &model.GetTripByIDRequest{}, trip.ID)
}

func (s *TripServiceImpl) UpdateTrip(ctx context.Context, id uuid.UUID, req *model.UpdateTripRequest) (*model.Trip, error) {
	trip, err := s.tripRepo.GetTripByID(ctx, &model.GetTripByIDRequest{}, id)
	if err != nil {
		return nil, ginext.NewInternalServerError("failed to get trip")
	}

	// Update fields if provided
	if req.DepartureTime != nil {
		if req.DepartureTime.Before(time.Now()) {
			return nil, ginext.NewBadRequestError("departure time cannot be in the past")
		}
		trip.DepartureTime = *req.DepartureTime
	}

	if req.ArrivalTime != nil {
		if req.ArrivalTime.Before(trip.DepartureTime) {
			return nil, ginext.NewBadRequestError("arrival time must be after departure time")
		}
		trip.ArrivalTime = *req.ArrivalTime
	}

	if req.BasePrice != nil {
		if *req.BasePrice < 0 {
			return nil, ginext.NewBadRequestError("base price must be non-negative")
		}
		trip.BasePrice = *req.BasePrice
	}

	if req.Status != nil {
		trip.Status = *req.Status
	}

	if req.IsActive != nil {
		trip.IsActive = *req.IsActive
	}

	if err := s.tripRepo.UpdateTrip(ctx, trip); err != nil {
		return nil, ginext.NewInternalServerError("failed to update trip")
	}

	return s.GetTripByID(ctx, &model.GetTripByIDRequest{}, id)
}

func (s *TripServiceImpl) DeleteTrip(ctx context.Context, id uuid.UUID) error {
	trip, err := s.tripRepo.GetTripByID(ctx, &model.GetTripByIDRequest{}, id)
	if err != nil {
		return fmt.Errorf("trip not found: %w", err)
	}

	if trip.Status != "scheduled" {
		return ginext.NewBadRequestError("only scheduled trips can be deleted")
	}

	if trip.DepartureTime.Before(time.Now().Add(24 * time.Hour)) {
		return ginext.NewBadRequestError("cannot delete trip within 24 hours of departure")
	}

	if err := s.tripRepo.DeleteTrip(ctx, id); err != nil {
		return ginext.NewInternalServerError("failed to delete trip")
	}

	return nil
}

// GetCompletedTripsForReschedule gets completed trips from last 24h that need rescheduling
func (s *TripServiceImpl) GetCompletedTripsForReschedule(ctx context.Context) ([]model.Trip, error) {
	return s.tripRepo.GetCompletedTripsForReschedule(ctx)
}

// RescheduleTrip updates trip with new departure/arrival times (maintains same bus, route, price)
func (s *TripServiceImpl) RescheduleTrip(ctx context.Context, id uuid.UUID, newDeparture, newArrival time.Time) error {
	trip, err := s.tripRepo.GetTripByID(ctx, &model.GetTripByIDRequest{}, id)
	if err != nil {
		return fmt.Errorf("trip not found: %w", err)
	}

	if newArrival.Before(newDeparture) {
		return ginext.NewBadRequestError("arrival time must be after departure time")
	}

	// Update times and reset status to scheduled
	trip.DepartureTime = newDeparture
	trip.ArrivalTime = newArrival
	trip.Status = "scheduled"

	if err := s.tripRepo.UpdateTrip(ctx, trip); err != nil {
		return fmt.Errorf("failed to reschedule trip: %w", err)
	}

	return nil
}

// ProcessTripStatusUpdates triggers the batch update of trip statuses
func (s *TripServiceImpl) ProcessTripStatusUpdates(ctx context.Context) error {
	return s.tripRepo.UpdateTripStatuses(ctx)
}

func (s *TripServiceImpl) CancelTrip(ctx context.Context, id uuid.UUID) error {
	// 1. Get Trip
	trip, err := s.tripRepo.GetTripByID(ctx, &model.GetTripByIDRequest{}, id)
	if err != nil {
		return ginext.NewInternalServerError("failed to get trip")
	}

	// 2. Validate Status Check
	// Only Scheduled or Delayed trips can be cancelled
	if trip.Status != constants.TripStatusScheduled && trip.Status != constants.TripStatusDelayed {
		return ginext.NewBadRequestError(fmt.Sprintf("Cannot cancel trip with status: %s. Only scheduled or delayed trips can be cancelled.", trip.Status))
	}

	// 3. Update Status to Cancelled
	trip.Status = constants.TripStatusCancelled
	
	if err := s.tripRepo.UpdateTrip(ctx, trip); err != nil {
		return ginext.NewInternalServerError("failed to update trip status")
	}

	// 4. Get Bookings
	bookings, err := s.bookingClient.GetTripBookings(ctx, id)
	if err != nil {
		// Log error but we already cancelled the trip in DB. 
		log.Error().Err(err).Str("trip_id", id.String()).Msg("Failed to get bookings for cancelled trip for processing refunds")
		return nil 
	}

	// 5. Refund/Cancel Bookings
	for _, b := range bookings {
		// Refund if PAID
		if b.TransactionStatus == "PAID" {
			req := &payment.RefundRequest{
				BookingID:    b.ID,
				Reason:       "Trip Cancelled by Operator",
				RefundAmount: b.TotalAmount,
			}
			_, err := s.paymentClient.CreateRefund(ctx, req)
			if err != nil {
				log.Error().Err(err).Str("booking_id", b.ID.String()).Msg("Failed to process refund for booking")
			}
		}

		// Cancel Booking
		err := s.bookingClient.CancelBooking(ctx, b.ID, "Trip Cancelled by Operator")
		if err != nil {
			log.Error().Err(err).Str("booking_id", b.ID.String()).Msg("Failed to cancel booking")
		}
	}

	return nil
}

File: frontend/lib/api/trip/trip-service.ts
import apiClient, { ApiResponse, handleApiError } from "../client";
import { getValue } from "@/lib/utils";
import {
  TripSearchParams,
  TripSearchResponse,
  ApiTripSearchResponse,
  Trip,
  SeatAvailabilityResponse,
  SeatDetail,
} from "@/lib/types/trip";

/**
 * Search trips
 */
export const searchTrips = async (
  params: TripSearchParams,
): Promise<TripSearchResponse> => {
  try {
    // The API returns the data directly, not wrapped in ApiResponse
    const response = await apiClient.get<ApiTripSearchResponse>(
      "/trip/api/v1/trips/search",
      { params },
    );

    const apiResponse = response.data;

    if (!apiResponse.data) {
      throw new Error("Failed to search trips: No data received");
    }

    // Return API response items directly without transformation
    // This preserves the route, bus structure for admin tables
    return {
      trips: apiResponse.data,
      total: apiResponse.meta.total,
      page: apiResponse.meta.page,
      page_size: apiResponse.meta.page_size,
      total_pages: apiResponse.meta.total_pages,
    };
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Get trip by ID
 */
export const getTripById = async (
  id: string,
  preload_route = true,
  preload_route_stop = false,
  preload_bus = true,
  preload_seat = false,
  seat_booking_status = false,
): Promise<Trip> => {
  try {
    // Build query params
    const params = new URLSearchParams();

    if (preload_route) {
      params.append("preload_route", "true");
    }
    if (preload_route_stop) {
      params.append("preload_route_stop", "true");
    }
    if (preload_bus) {
      params.append("preload_bus", "true");
    }
    if (preload_seat) {
      params.append("preload_seat", "true");
    }
    if (seat_booking_status) {
      params.append("seat_booking_status", "true");
    }

    const queryString = params.toString();

    const response = await apiClient.get<{ data: Trip }>(
      `/trip/api/v1/trips/${id}${queryString ? `?${queryString}` : ""}`,
    );

    if (!response.data.data) {
      throw new Error("Failed to get trip: No data received");
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Get seat availability for a trip
 */
export const getTripSeats = async (
  tripId: string,
): Promise<SeatAvailabilityResponse> => {
  try {
    // The separate seats endpoint is deprecated/404.
    // We fetch the full trip details which includes bus.seats with availability status.
    const trip = await getTripById(tripId);

    if (!trip.bus || !trip.bus.seats) {
      throw new Error("Bus seat information not available for this trip");
    }

    const seats: SeatDetail[] = trip.bus.seats.map((seat) => ({
      id: seat.id,
      seat_code: seat.seat_number,
      seat_type: getValue(seat.seat_type), // Extract value from ConstantDisplay
      is_booked: !seat.is_available, // In the new API, is_available=true means NOT booked
      is_locked: false, // No locking info in current API response
      price: trip.base_price * seat.price_multiplier,
    }));

    const availableSeats = seats.filter((s) => !s.is_booked).length;

    return {
      trip_id: trip.id,
      available_seats: availableSeats,
      total_seats: trip.bus.seat_capacity,
      seats,
    };
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Create a new trip (admin only)
 */
export const createTrip = async (tripData: {
  route_id: string;
  bus_id: string;
  departure_time: string; // ISO datetime string
  arrival_time: string; // ISO datetime string
  base_price: number;
}): Promise<Trip> => {
  try {
    const response = await apiClient.post<ApiResponse<Trip>>(
      "/trip/api/v1/trips",
      tripData,
    );

    if (!response.data.data) {
      throw new Error("Failed to create trip");
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Update a trip (admin only)
 */
export const updateTrip = async (
  id: string,
  tripData: {
    departure_time?: string;
    arrival_time?: string;
    base_price?: number;
    status?: string;
    is_active?: boolean;
  },
): Promise<Trip> => {
  try {
    const response = await apiClient.put<ApiResponse<Trip>>(
      `/trip/api/v1/trips/${id}`,
      tripData,
    );

    if (!response.data.data) {
      throw new Error("Failed to update trip");
    }

    return response.data.data;
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * Delete a trip (admin only)
 */
export const deleteTrip = async (id: string): Promise<void> => {
  try {
    await apiClient.delete(`/trip/api/v1/trips/${id}`);
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};

/**
 * List trips with pagination (admin only)
 */
export const listTrips = async (params?: {
  page?: number;
  page_size?: number;
  search?: string;
  status?: string;
}): Promise<{
  trips: Trip[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}> => {
  try {
    // The backend returns { data: Trip[], meta: { page, page_size, total, total_pages } }
    // We need to map this to our internal return type
    const response = await apiClient.get<
      ApiResponse<Trip[]> & {
        meta: {
          page: number;
          page_size: number;
          total: number;
          total_pages: number;
        };
      }
    >("/trip/api/v1/trips", {
      params: {
        page: params?.page || 1,
        page_size: params?.page_size || 5,
        search: params?.search,
        status: params?.status,
      },
    });

    if (!response.data.data) {
      // If data is null/undefined, return empty list
      return {
        trips: [],
        total: 0,
        page: params?.page || 1,
        page_size: params?.page_size || 5,
        total_pages: 0,
      };
    }

    return {
      trips: response.data.data,
      total: response.data.meta?.total || 0,
      page: response.data.meta?.page || 1,
      page_size: response.data.meta?.page_size || 5,
      total_pages: response.data.meta?.total_pages || 1,
    };
  } catch (error) {
    const errorMessage = handleApiError(error);
    throw new Error(errorMessage);
  }
};
